src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5573,6 +5573,307 @@ class StubGenerator: public StubCodeGenerator {
+      255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,  62u, 255u, 255u,
+       52u,  53u,  54u,  55u,  56u,  57u,  58u,  59u,  60u,  61u, 255u, 255u, 255u, 255u, 255u, 255u,
+        0u, 255u,   0u,   1u,   2u,   3u,   4u,   5u,   6u,   7u,   8u,   9u,  10u,  11u,  12u,  13u,
+       14u,  15u,  16u,  17u,  18u,  19u,  20u,  21u,  22u,  23u,  24u,  25u, 255u, 255u, 255u, 255u,
+       63u, 255u,  26u,  27u,  28u,  29u,  30u,  31u,  32u,  33u,  34u,  35u,  36u,  37u,  38u,  39u,
+       40u,  41u,  42u,  43u,  44u,  45u,  46u,  47u,  48u,  49u,  50u,  51u, 255u, 255u, 255u, 255u,
+    };
+
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "decodeBlock");
+    address start = __ pc();
+
+    Register src   = c_rarg0;  // source array
+    Register soff  = c_rarg1;  // source start offset
+    Register send  = c_rarg2;  // source end offset
+    Register dst   = c_rarg3;  // dest array
+    Register doff  = c_rarg4;  // position for writing to dest array
+    Register isURL = c_rarg5;  // Base64 or URL character set
+
+    Register length = send;    // reuse send as length of source data to process
+
+    Register simd_codec   = c_rarg6;
+    Register nosimd_codec = c_rarg7;
+
+    Label ProcessData, Process64B, Process32B, Process4B, SIMDEnter, SIMDExit, Exit;
+
+    __ enter();
+
+    __ add(src, src, soff);
+    __ add(dst, dst, doff);
+
+    __ mov(doff, dst);
+
+    __ sub(length, send, soff);
+    __ bfm(length, zr, 0, 1);
+
+    __ lea(nosimd_codec, ExternalAddress((address) fromBase64ForNoSIMD));
+    __ cbz(isURL, ProcessData);
+    __ lea(nosimd_codec, ExternalAddress((address) fromBase64URLForNoSIMD));
+
+    __ BIND(ProcessData);
+    __ mov(rscratch1, length);
+    __ cmp(length, (u1)144); // 144 = 80 + 64
+    __ br(Assembler::LT, Process4B);
+
+    // In the MIME case, the line length cannot be more than 76
+    // bytes (see RFC 2045). This is too short a block for SIMD
+    // to be worthwhile, so we use non-SIMD here.
+    __ movw(rscratch1, 79);
+
+    __ BIND(Process4B);
+    __ ldrw(r14, __ post(src, 4));
+    __ ubfxw(r10, r14, 0,  8);
+    __ ubfxw(r11, r14, 8,  8);
+    __ ubfxw(r12, r14, 16, 8);
+    __ ubfxw(r13, r14, 24, 8);
+    // get the de-code
+    __ ldrb(r10, Address(nosimd_codec, r10, Address::uxtw(0)));
+    __ ldrb(r11, Address(nosimd_codec, r11, Address::uxtw(0)));
+    __ ldrb(r12, Address(nosimd_codec, r12, Address::uxtw(0)));
+    __ ldrb(r13, Address(nosimd_codec, r13, Address::uxtw(0)));
+    // error detection, 255u indicates an illegal input
+    __ orrw(r14, r10, r11);
+    __ orrw(r15, r12, r13);
+    __ orrw(r14, r14, r15);
+    __ tbnz(r14, 7, Exit);
+    // recover the data
+    __ lslw(r14, r10, 2);
+    __ bfmw(r14, r11, 4, 5);
+    __ lslw(r15, r11, 4);
+    __ bfmw(r15, r12, 2, 5);
+    __ orrw(r13, r13, r12, Assembler::LSL, 6);
+    __ strb(r14, __ post(dst, 1));
+    __ strb(r15, __ post(dst, 1));
+    __ strb(r13, __ post(dst, 1));
```

Comment:
I think this sequence should be 4 BFMs, STRW, BFM, STRW. That's the best we can do, I think.

Reply:
Sorry, that's not quite right, but you get the idea: let's not generate unnecessary memory traffic.

Reply:
Okay, implemented as:
```
__ lslw(r14, r10, 10);
__ bfiw(r14, r11, 4, 6);
__ bfmw(r14, r12, 2, 5);
__ rev16w(r14, r14);
__ bfiw(r13, r12, 6, 2);
__ strh(r14, __ post(dst, 2));
__ strb(r13, __ post(dst, 1));
```